### PR TITLE
Support EV route refresh

### PR DESCRIPTION
**Pull request by dzinad**
### Description
The PR has the following parts (split into the corresponding commits):
1. Get EV data updates from client and pass it as route refresh request parameters;
2. Update state_of_charge annotations after route refresh (provided via unrecognized properties);
3. Update waypoints after route refresh.

### CODE REVIEW

File: instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/EVRouteRefreshTest.kt
Comment (dzinad) [inline instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/EVRouteRefreshTest.kt:506]: This check fails for now because I've used the real Directions API response to form our test response but it returned less waypoints than I expected (and than as described in their docs). I've asked Mandeep a corresponding question.

**Comment by dzinad** [comment]
I've added a DO NOT MERGE label because some changes may be required after #6005 is merged.

File: libnavigation-base/src/main/java/com/mapbox/navigation/base/internal/RouteRefreshRequestData.kt
Comment (LukasPaczos) [inline libnavigation-base/src/main/java/com/mapbox/navigation/base/internal/RouteRefreshRequestData.kt:19]:
```suggestion
    val experimentalProperties: Map<String, String>,
```
nit

File: libnavigation-base/src/main/java/com/mapbox/navigation/base/internal/route/AnnotationsRefresher.kt
Comment (LukasPaczos) [inline libnavigation-base/src/main/java/com/mapbox/navigation/base/internal/route/AnnotationsRefresher.kt:55]: Why do we discard the new data if original response was empty? If we have new information we should add it to the existing instance.

File: libnavigation-base/src/main/java/com/mapbox/navigation/base/internal/route/AnnotationsRefresher.kt
Comment (LukasPaczos) [inline libnavigation-base/src/main/java/com/mapbox/navigation/base/internal/route/AnnotationsRefresher.kt:111]:
```suggestion
        return JsonArray(this.size).also { array ->
```
small improvement that optimizes unnecessary array resizes (since we know the resulting size)

File: libnavigation-base/src/main/java/com/mapbox/navigation/base/internal/route/NavigationRouteEx.kt
Comment (LukasPaczos) [inline libnavigation-base/src/main/java/com/mapbox/navigation/base/internal/route/NavigationRouteEx.kt:110]:
```suggestion
        directionsRouteBlock = {
            toBuilder()
                .legs(updateLegs)
                .updateRouteDurationBasedOnLegsDuration(updateLegs)
                .build()
        },
        directionsResponseBlock = {
            updateWaypoints(
                directionsResponse.waypoints(),
                unrecognizedProperties
            )
        }
    )
```
nit, to improve readability

File: libnavigation-base/src/main/java/com/mapbox/navigation/base/internal/route/NavigationRouteEx.kt
Comment (LukasPaczos) [inline libnavigation-base/src/main/java/com/mapbox/navigation/base/internal/route/NavigationRouteEx.kt:49]:
```suggestion
    waypoints: List<DirectionsWaypoint>?,
```
This makes more sense from the perspective that anything could be refreshed via unregonized properties in the future and passing just the map gives no indication to this function what in the response should that map actually impact.

File: libnavigation-base/src/main/java/com/mapbox/navigation/base/internal/route/NavigationRouteEx.kt
Comment (LukasPaczos) [inline libnavigation-base/src/main/java/com/mapbox/navigation/base/internal/route/NavigationRouteEx.kt:208]: Great if this could happen in `RouterWrapper` as suggested above.

File: libnavigation-base/src/main/java/com/mapbox/navigation/base/internal/route/AnnotationsRefresher.kt
Comment (dzinad) [reply libnavigation-base/src/main/java/com/mapbox/navigation/base/internal/route/AnnotationsRefresher.kt:55]: It's an annotation that's supposed to be refreshed using some quite strict rules. If we used current_route_geometry_index, old values are not returned in the refresh response. The response may be truncated so we have to fill the rest with the old values. We could use some default values but for the other annotations we require the original ones to be filled. Even for this case: if we fill everything that's not returned in the refresh response with 0 or 100, it may break some user algorithms based on these annotations.

Comment (LukasPaczos) [reply libnavigation-base/src/main/java/com/mapbox/navigation/base/internal/route/AnnotationsRefresher.kt:55]: That makes sense :+1:

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/MapboxNavigation.kt
Comment (LukasPaczos) [inline libnavigation-core/src/main/java/com/mapbox/navigation/core/MapboxNavigation.kt:1678]:
```suggestion
    fun onEVDataUpdated(data: Map<String, String?>) {
```
Why not simplify like this? What's the benefit of forcing the developers to implement the full observer pattern?

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/routerefresh/RouteRefreshController.kt
Comment (LukasPaczos) [inline libnavigation-core/src/main/java/com/mapbox/navigation/core/routerefresh/RouteRefreshController.kt:184]: nit: named parameters would help with readability.

File: libnavigation-router/src/main/java/com/mapbox/navigation/route/internal/WaypointsParser.kt
Comment (LukasPaczos) [inline libnavigation-router/src/main/java/com/mapbox/navigation/route/internal/WaypointsParser.kt:21]: Could you add an error log?

File: libnavigator/src/main/java/com/mapbox/navigation/navigator/internal/MapboxNativeNavigatorImpl.kt
Comment (LukasPaczos) [inline libnavigator/src/main/java/com/mapbox/navigation/navigator/internal/MapboxNativeNavigatorImpl.kt:271]: Good to remove?

File: libnavigation-base/src/main/java/com/mapbox/navigation/base/internal/utils/Constants.java
Comment (dzinad) [reply libnavigation-base/src/main/java/com/mapbox/navigation/base/internal/utils/Constants.java:5]: I did a bit differently.

File: libnavigator/src/main/java/com/mapbox/navigation/navigator/internal/MapboxNativeNavigatorImpl.kt
Comment (dzinad) [reply libnavigator/src/main/java/com/mapbox/navigation/navigator/internal/MapboxNativeNavigatorImpl.kt:271]: Argh, I removed the log but left this loop...

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/MapboxNavigation.kt
Comment (dzinad) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/MapboxNavigation.kt:1678]: I guess the initial purpose was to isolate the logic from `MapboxNavigation`. In case it became more complicated. So that I don't have to add 5 ev related methods to `MapboxNavigation`.
But it does require more actions from the user and it's only 1 method so I guess it makes sense to simplify.

File: instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/EVRouteRefreshTest.kt
Comment (dzinad) [reply instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/EVRouteRefreshTest.kt]: The fixed a bug, I've updated the mocked response appropriately. Should be working now.

**Comment by LukasPaczos** [comment]
Some minor comments above and let's wait for https://github.com/mapbox/mapbox-navigation-android/pull/6005 (if we need to) but in general LGTM!

File: libnavigation-base/src/main/java/com/mapbox/navigation/base/internal/utils/Constants.java
Comment (dzinad) [reply libnavigation-base/src/main/java/com/mapbox/navigation/base/internal/utils/Constants.java:8]: Because in kotlin this code:
```
object ConstantsKt {
    object RouteResponse {
        const val KEY_WAYPOINTS = "waypoints"
    }
}
```
looks like this:
```
public final class ConstantsKt {
   @NotNull
   public static final ConstantsKt INSTANCE;

   private ConstantsKt() {
   }

   static {
      ConstantsKt var0 = new ConstantsKt();
      INSTANCE = var0;
   }

   @Metadata(
      mv = {1, 5, 1},
      k = 1,
      d1 = {"\u0000\u0012\n\u0002\u0018\u0002\n\u0002\u0010\u0000\n\u0002\b\u0002\n\u0002\u0010\u000e\n\u0000\bÆ\u0002\u0018\u00002\u00020\u0001B\u0007\b\u0002¢\u0006\u0002\u0010\u0002R\u000e\u0010\u0003\u001a\u00020\u0004X\u0086T¢\u0006\u0002\n\u0000¨\u0006\u0005"},
      d2 = {"Lcom/mapbox/navigation/base/internal/utils/ConstantsKt$RouteResponse;", "", "()V", "KEY_WAYPOINTS", "", "mapbox-navigation-android.libnavigation-base.main"}
   )
   public static final class RouteResponse {
      @NotNull
      public static final String KEY_WAYPOINTS = "waypoints";
      @NotNull
      public static final ConstantsKt.RouteResponse INSTANCE;

      private RouteResponse() {
      }

      static {
         ConstantsKt.RouteResponse var0 = new ConstantsKt.RouteResponse();
         INSTANCE = var0;
      }
   }
}
```
Isn't it too much for a string constant?
I can add the `JvmStatic` annotation, but it's not much better.
Do you have anything against java? I can easily convert it to kotlin but I think that for the purposes like this (when you just need a holder for some constants) java is better (probably the only situation where I'd say this).

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/MapboxNavigation.kt
Comment (dzinad) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/MapboxNavigation.kt:1670]: It's not a problem but I want to specify that it's not necessary to implement the caching on the app side. Changed the docs.

**Comment by dzinad** [comment]
> Some minor comments above and let's wait for https://github.com/mapbox/mapbox-navigation-android/pull/6005 (if we need to)

Whoever merges first will leave the second person supporting these changes. :)

File: libnavigation-base/src/main/java/com/mapbox/navigation/base/internal/utils/Constants.java
Comment (dzinad) [reply libnavigation-base/src/main/java/com/mapbox/navigation/base/internal/utils/Constants.java]: Converted to kotlin.

Comment (LukasPaczos) [reply libnavigation-base/src/main/java/com/mapbox/navigation/base/internal/utils/Constants.java]: Thanks, that's an interesting point of view. I guess it's a general problem with `object`s that they are just singletons under the hood after compilation. Either way would work, I was just curious what was the motivation.

**Review by LukasPaczos — APPROVED** [review]

**Comment by RingerJK** [comment]
https://github.com/mapbox/mapbox-navigation-android/pull/6005 has landed

**Comment by Mundis2201** [comment]
`CHANGELOG.md`

**Pull request merged**
